### ISSUE TITLE
GPU + openacc broken for odd number of GPUs on same node

### ISSUE THREAD

**Issue by jkwashbourne** [devitocodes/devito#1655]
Devito with variables `DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc DEVITO_LANGUAGE=openacc DEVITO_MPI=1` only works for certain numbers of MPI ranks. On a machine with 8 GPUs, only the numbers [1,2,4,8] work.

### Test platforms: 
* azure `ND40rs_v2` with 8 v100 GPUs
    * working number of MPI ranks: [1, 2, 4, 8] 
* azure `NC24_v3` with 4 v100 GPUs  
    * working number of MPI ranks: [1, 2, 4] 

### Reproducer
`DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 3 python3 ./examples/seismic/acoustic/acoustic_example.py 
`

### Output log from command
```
Operator `initdamp` ran in 6.08 s
Operator `initdamp` ran in 6.08 s
Operator `initdamp` ran in 6.08 s
Operator `pad_vp` ran in 0.01 s
Operator `pad_vp` ran in 0.01 s
Operator `pad_vp` ran in 0.01 s
Applying Forward
Applying Forward
Applying Forward
hostptr=0x2426ce0,stride=1,size=3,extent=-3,eltsize=4,name=src_coords,flags=0x200=present,async=-1,threadid=1
Present table dump for device[1]: NVIDIA Tesla GPU 0, compute capability 7.0, threadid=1
...empty...
deleted block   device:0x7eff17800000 size:3184128 threadid=1 
deleted block   device:0x7eff17c00000 size:3941888 threadid=1 
Present table dump for device[3]: NVIDIA Tesla GPU 2, compute capability 7.0, threadid=1
host:0x205a8e0 device:0x7efef5b7a800 size:2448 presentcount:0+1 line:76 name:rec_coords
host:0x2426ce0 device:(nil) size:0 presentcount:0+1 line:76 name:src_coords
host:0x317f000 device:0x7efee1000000 size:3941484 presentcount:0+1 line:76 name:vp
host:0x36c1e40 device:(nil) size:0 presentcount:1+1 line:76 name:src
host:0x37e8000 device:0x7efef5afa000 size:526320 presentcount:0+1 line:76 name:rec
host:0x457e000 device:0x7efee0000000 size:11824452 presentcount:0+1 line:76 name:u
host:0x7eff7ffd7000 device:0x7efee0c00000 size:3184020 presentcount:0+1 line:76 name:damp
allocated block device:0x7efee0000000 size:11824640 thread:1
allocated block device:0x7efee0c00000 size:3184128 thread:1
allocated block device:0x7efee1000000 size:3941888 thread:1
allocated block device:0x7efef5afa000 size:526336 thread:1
allocated block device:0x7efef5b7a800 size:2560 thread:1
deleted block   device:0x7efef5bc6a00 size:113664 threadid=1 
deleted block   device:0x7efef5be2600 size:113664 threadid=1 
deleted block   device:0x7efef5b7b200 size:309248 threadid=1 
deleted block   device:0x7efee1400000 size:309248 threadid=1 

 file:/scratch/devito-jitcache-uid1000/6e5a46e9897efcd1ef003ac02a492d6bb8d67074.cpp Forward line:98

hostptr=0x21f11f0,stride=1,size=3,extent=-3,eltsize=4,name=src_coords,flags=0x200=present,async=-1,threadid=1
Present table dump for device[1]: NVIDIA Tesla GPU 0, compute capability 7.0, threadid=1
host:0x21f11f0 device:(nil) size:0 presentcount:0+1 line:76 name:src_coords
host:0x2f6f000 device:0x7f381fc00000 size:4018768 presentcount:0+1 line:76 name:vp
host:0x377cb90 device:0x7f3815b7a800 size:2448 presentcount:0+1 line:76 name:rec_coords
host:0x37a2240 device:(nil) size:0 presentcount:1+1 line:76 name:src
host:0x3ede810 device:0x7f3815afa000 size:526320 presentcount:0+1 line:76 name:rec
host:0x4585000 device:0x7f3800000000 size:12056304 presentcount:0+1 line:76 name:u
host:0x7f3885383000 device:0x7f381f800000 size:3254776 presentcount:0+1 line:76 name:damp
allocated block device:0x7f3800000000 size:12056576 thread:1
allocated block device:0x7f3815afa000 size:526336 thread:1
allocated block device:0x7f3815b7a800 size:2560 thread:1
allocated block device:0x7f381f800000 size:3254784 thread:1
allocated block device:0x7f381fc00000 size:4019200 thread:1
deleted block   device:0x7f3815bc6a00 size:115712 threadid=1 
deleted block   device:0x7f3815be2e00 size:115712 threadid=1 
deleted block   device:0x7f3815b7b200 size:309248 threadid=1 
deleted block   device:0x7f3800c00000 size:309248 threadid=1 

 file:/scratch/devito-jitcache-uid1000/6e5a46e9897efcd1ef003ac02a492d6bb8d67074.cpp Forward line:98

--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[56704,1],0]
  Exit code:    1
--------------------------------------------------------------------------
```

**Comment by FabioLuporini** [comment]
@jkwashbourne I think I may know what's going on here. With certain numbers of MPI ranks, you get eg 0 sources on some ranks, which means the device pointer pointing to `src` is NULL. So, when the runtime attempts to dereference it in an OpenACC parloop, it crashes. Will investigate and let u know

**Comment by jkwashbourne** [comment]
I thought it was unlikely that for example 8 works but 7 does not, or 4 works but 3 and 5 do not. but maybe ...

**Comment by FabioLuporini** [comment]
yeah, but might be that it's the odd numbers that create such MPI ranks with zero soruces.

I've just merged a patch. Could you test again?

**Comment by jkwashbourne** [comment]
closed by #1684

The most recent PR closes this issue. Here are results for 1201,1201,601 skew adjoint isotropic with `openacc` offloading. 

```
out.openacc.mpi0.txt Global performance: [OI=2.17,  217.14 GFlops/s,  2.27 GPts/s]
out.openacc.mpi1.txt Global performance: [OI=2.17,  217.11 GFlops/s,  2.27 GPts/s]
out.openacc.mpi2.txt Global performance: [OI=2.17,  482.34 GFlops/s,  5.03 GPts/s]
out.openacc.mpi4.txt Global performance: [OI=2.17, 1017.35 GFlops/s, 10.58 GPts/s]
out.openacc.mpi5.txt Global performance: [OI=2.17, 1362.43 GFlops/s, 14.12 GPts/s]
out.openacc.mpi6.txt Global performance: [OI=2.17, 1430.79 GFlops/s, 14.85 GPts/s]
out.openacc.mpi7.txt Global performance: [OI=2.17, 1859.55 GFlops/s, 19.19 GPts/s]
out.openacc.mpi8.txt Global performance: [OI=2.17, 1925.80 GFlops/s, 19.95 GPts/s]
```

commands:
```
export nt=500
export nx=1201
export ny=1201
export nz=601

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=0 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi0.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 1 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi1.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 2 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi2.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 3 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi4.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 4 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi4.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 5 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi5.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 6 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi6.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 7 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi7.txt

DEVITO_JIT_BACKDOOR=0 DEVITO_LOGGING=DEBUG DEVITO_PLATFORM=nvidiaX DEVITO_ARCH=nvc \
DEVITO_LANGUAGE=openacc DEVITO_MPI=1 mpirun -n 8 python3 example_iso.py $nt $nx $ny $nz 0 >& out.openacc.mpi8.txt
```

**Comment by jkwashbourne** [comment]
Great work devito team!
